README.md
# Check\_Graphite

## Usage

### check target against thresholds
```
./check_graphite -u http://localhost:8080 -target=collectd.host1.memory.percent-used -c 90 -w 80
```

### calculate percentage in pairs of targets and check percentage against thresholds
```
./check_graphite -u http://localhost:8080 -target=collectd.host1.GenericJMX-app1_memory-heap.memory-{used,max} -target=collectd.host1.GenericJMX-app1_memory_pool-CMS_Perm_Gen.memory-{used,max} -p -c 42 -w 2 -r "memory[^-]*-(.*)$"
```

### check sum of targets against thresholds
```
./check_graphite -u http://localhost:8080 -target=collectd.host1.aggregation-cpu-average.cpu-{system,user} -w 70 -c 90 -s -name cpu-usage
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/lscheidler/check_graphite.

## License

The code is available as open source under the terms of the [Apache 2.0 License](http://opensource.org/licenses/Apache-2.0).


CHANGELOG.md
# 0.1.5 (2021-03-17)

- added go.mod
- updated go-nagios module

# 0.1.4 (2020-08-28)

- refactored code
- used go fmt to format code
- fix golint warnings
- added -debug argument

# 0.1.3 (2018-05-29)

- added Version constant with plugin version
- bump go-nagios to 0.1.1

# 0.1.2 (2018-05-29)

- added -empty-ok flag, handle empty datapoints as critical by default

# 0.1.1 (2018-05-29)

- support globs in targets

# 0.1.0 (2018-05-29)

- Initial release


graphite/utils_test.go
/*
Copyright 2018 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package graphite provides struct and functions to get datapoints from graphite web
package graphite

import (
	"encoding/json"
	"testing"
)

func TestParseJSONData(t *testing.T) {
	var msg = "[{\"target\": \"test\", \"datapoints\": [[3.5805225, 1526986610], [3.5805225, 1526986611]]}]"
	var jsonData interface{}
	unmarshalErr := json.Unmarshal([]byte(msg), &jsonData)
	if unmarshalErr != nil {
		t.Error(
			"Unmarshalling failed for", msg,
		)
	}

	var targetsData = ParseJSONData(jsonData)
	if len(targetsData) == 0 {
		t.Error(
			"Should not empty", targetsData,
		)
	}

	if len(targetsData["test"].datapoints) == 0 {
		t.Error(
			"Should not empty", targetsData["test"].datapoints,
		)
	}
}

func TestAverage(t *testing.T) {
	var msg = "[{\"target\": \"test\", \"datapoints\": [[3.5805225, 1526986610], [3.5805225, 1526986611]]}, {\"target\": \"test2\", \"datapoints\": [[3, 1526986610], [6, 1526986611], [3, 1526986612]]}]"
	var jsonData interface{}
	json.Unmarshal([]byte(msg), &jsonData)
	var targetsData = ParseJSONData(jsonData)

	if targetsData["test"].Average() != 3.5805225 {
		t.Error(
			"Should be average", targetsData["test"].Average(),
		)
	}

	if targetsData["test2"].Average() != 4 {
		t.Error(
			"Should be average", targetsData["test2"].Average(),
		)
	}
}


main.go
/*
Copyright 2021 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package main contains check_graphite
package main

import (
	"github.com/lscheidler/check_graphite/check"
)

const version = "0.1.5"

// main initialize nagios struct and run check
func main() {
	checkGraphite := check.Init(version)
	checkGraphite.Run()
}


check/check.go
/*
Copyright 2021 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package check provides functions regarding graphite check
package check

import (
	"fmt"
	"math"
	"regexp"
	"strings"

	"github.com/lscheidler/check_graphite/graphite"
	"github.com/lscheidler/go-nagios"
)

// Check data structure
type Check struct {
	Critical         float64
	Debug            bool
	EmptyOk          bool
	Name             string
	nagios           *nagios.Nagios
	Max              float64
	Percentage       bool
	Perfdata         bool
	SumFlag          bool
	TargetFlag       *target
	TargetNameRegexp string
	Warning          float64
}

type target []string

// String is the method to format the flag's value, part of the flag.Value interface.
// The String method's output will be used in diagnostics.
func (i *target) String() string {
	return fmt.Sprint(*i)
}

// Set is the method to set the flag value, part of the flag.Value interface.
// Set's argument is a string to be parsed to set the flag.
// It's a comma-separated list, so we split it.
func (i *target) Set(value string) error {
	for _, t := range strings.Split(value, ",") {
		*i = append(*i, t)
	}
	return nil
}

// Exit defer function
func (check *Check) Exit() {
	check.nagios.Exit()
}

func (check *Check) Error(err error) {
	check.nagios.Unknown(err.Error())
}

// check checks graphite targets against thresholds
func (check *Check) runCheck(data map[string]graphite.Target) {
	if check.Debug {
		fmt.Printf("%v\n", data)
	}

	check.nagios.ShowPerfdata = check.Perfdata

	if check.Percentage && math.IsNaN(check.Max) {
		check.checkPercentage(data)
	} else {
		sum := check.checkTargets(data)

		if check.SumFlag {
			if check.Percentage && !math.IsNaN(check.Max) {
				check.nagios.CheckPercentageThreshold(check.Name, sum, check.Max, check.Warning, check.Critical)
			} else {
				check.nagios.CheckThreshold(check.Name, sum, check.Warning, check.Critical)
			}
		}
	}
}

// checkPercentage checks the percentage of graphite targets against thresholds
func (check *Check) checkPercentage(data map[string]graphite.Target) {
	if len(data)%2 == 0 {
		for i := 0; i < len(*check.TargetFlag); i += 2 {
			targetValue, targetValueOk := graphite.Find(data, (*check.TargetFlag)[i])
			targetValueOk = check.checkTargetValue(targetValue, targetValueOk, (*check.TargetFlag)[i])

			targetMax, targetMaxOk := graphite.Find(data, (*check.TargetFlag)[i+1])
			targetMaxOk = check.checkTargetValue(targetMax, targetMaxOk, (*check.TargetFlag)[i+1])

			if targetValueOk && targetMaxOk {
				check.checkPercentagePair(targetValue, targetMax)
			}
		}
	} else {
		check.nagios.Unknown("Count of targets is not even, can not calculate percentage for targets.")
	}
}

func (check *Check) checkTargetValue(target graphite.Target, ok bool, targetName string) bool {
	result := ok
	if !ok {
		check.nagios.Unknown("target not found: " + targetName)
	}

	if target.IsEmpty() {
		result = false

		if check.EmptyOk {
			check.nagios.Ok("target is empty: " + target.Target())
		} else {
			check.nagios.Critical("target is empty: " + target.Target())
		}
	}
	return result
}

// checkPercentagePair checks the percentage of a pair of graphite targets against thresholds
func (check *Check) checkPercentagePair(value graphite.Target, max graphite.Target) {
	name := check.getName(value.Target())
	check.nagios.CheckPercentageThreshold(name, float64(value.Average()), float64(max.Average()), check.Warning, check.Critical)
}

// checkTargets checks targets against thresholds
func (check *Check) checkTargets(data map[string]graphite.Target) float64 {
	var sum float64
	for _, target := range *check.TargetFlag {
		if val, ok := data[target]; ok {
			if check.Debug {
				fmt.Printf("[%s] %v\n", target, val)
			}

			if check.SumFlag {
				sum += float64(val.Average())
				check.Name = check.getName(target)
			} else {
				name := check.getName(target)
				check.nagios.CheckThreshold(name, float64(val.Average()), check.Warning, check.Critical)
			}
		} else {
			check.nagios.Unknown("target not found: " + target)
		}
	}
	return sum
}

func (check *Check) getName(target string) string {
	var name string
	if name = check.Name; check.Name == "" {
		if name = target; check.TargetNameRegexp != "" {
			pattern := regexp.MustCompile(check.TargetNameRegexp)
			if name = pattern.FindString(target); len(pattern.FindStringSubmatch(target)) > 1 {
				matches := pattern.FindStringSubmatch(target)
				name = strings.Join(matches[1:], ".")
			}
		}
	}
	return name
}


graphite/utils.go
/*
Copyright 2018 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package graphite provides struct and functions to get datapoints from graphite web
package graphite

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"path/filepath"
	"strings"
	"time"
)

// Datapoint holds a graphite datapoint returned from graphite web
type Datapoint struct {
	value     float32
	timestamp int
}

// Target holds a target and its datapoints returned from graphite web
type Target struct {
	target     string
	datapoints []Datapoint
}

// Average returns the average calculated over all datapoints
func (target Target) Average() float32 {
	var sum float32
	for _, datapoint := range target.datapoints {
		sum += datapoint.value
	}
	return sum / float32(len(target.datapoints))
}

// IsEmpty returns true if datapoints slice is empty
func (target Target) IsEmpty() bool {
	return len(target.datapoints) == 0
}

// Target returns target name
func (target Target) Target() string {
	return target.target
}

// Find finds a target in data map
func Find(data map[string]Target, targetName string) (Target, bool) {
	target := Target{}
	targetOk := false

	// if targetName contains a glob, we need to match the glob with all targets available
	if strings.Contains(targetName, "*") {
		for _, targetElem := range data {
			if matched, _ := filepath.Match(targetName, targetElem.Target()); matched {
				return targetElem, true
			}
		}
	} else {
		target, targetOk = data[targetName]
	}

	return target, targetOk
}

// GetTargets gets data from graphite web and returns a slice of targets
func GetTargets(graphiteURL string, targets []string, from string, until string, timeout time.Duration) (map[string]Target, error) {
	jsonData, err := Get(graphiteURL, targets, from, until, timeout)
	if err != nil {
		return nil, err
	}

	return ParseJSONData(jsonData), nil
}

// ParseJSONData parses json data
func ParseJSONData(jsonData interface{}) map[string]Target {
	targetsData := map[string]Target{}
	for _, elem := range jsonData.([]interface{}) {
		elem := elem.(map[string]interface{})
		target := Target{target: elem["target"].(string), datapoints: []Datapoint{}}

		for _, datapointElem := range elem["datapoints"].([]interface{}) {
			datapointElem := datapointElem.([]interface{})

			// ignore datapoints, where value is nil
			if datapointElem[0] != nil {
				datapoint := Datapoint{timestamp: int(datapointElem[1].(float64)), value: float32(datapointElem[0].(float64))}

				target.datapoints = append(target.datapoints, datapoint)
			}
		}

		targetsData[target.target] = target
	}
	return targetsData
}

// Get gets datapoints from graphite web
func Get(graphiteURL string, targets []string, from string, until string, timeout time.Duration) (interface{}, error) {
	client := &http.Client{
		Timeout: time.Duration(timeout * time.Second),
	}

	resp, getErr := client.Get(graphiteURL + "/render?format=json&target=" + strings.Join(targets, "&target=") + "&from=" + from + "&until=" + until)
	if getErr != nil {
		return nil, getErr
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("Wrong status code: %d", resp.StatusCode)
	}

	body, bodyReadErr := ioutil.ReadAll(resp.Body)
	if bodyReadErr != nil {
		return nil, bodyReadErr
	}

	var jsonData interface{}
	unmarshalErr := json.Unmarshal(body, &jsonData)
	if unmarshalErr != nil {
		return nil, unmarshalErr
	}
	return jsonData, nil
}


check/cli.go
/*
Copyright 2021 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package check provides functions regarding graphite check
package check

import (
	"flag"
	"fmt"
	"math"
	"os"
	"time"

	"github.com/lscheidler/check_graphite/graphite"
	"github.com/lscheidler/go-nagios"
)

// flag variables
var (
	from        string
	graphiteURL string
	targetFlag  target
	timeout     time.Duration
	until       string
)

// Init initializes command line interface
func Init(version string) *Check {
	checkGraphite := Check{
		nagios: nagios.New(),
	}

	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "Usage of %s (%s):\n", os.Args[0], version)
		flag.PrintDefaults()

		examples := `
examples:
  # check target against thresholds
  ` + os.Args[0] + ` -u http://localhost:8080 -target=collectd.host1.memory.percent-used -c 90 -w 80

  # calculate percentage in pairs of targets and check percentage against thresholds
  ` + os.Args[0] + ` -u http://localhost:8080 -target=collectd.host1.GenericJMX-app1_memory-heap.memory-{used,max} -target=collectd.host1.GenericJMX-app1_memory_pool-CMS_Perm_Gen.memory-{used,max} -p -c 42 -w 2 -r "memory[^-]*-(.*)$"

  # check sum of targets against thresholds
  ` + os.Args[0] + ` -u http://localhost:8080 -target=collectd.host1.aggregation-cpu-average.cpu-{system,user} -w 70 -c 90 -s -name cpu-usage`

		fmt.Fprintf(flag.CommandLine.Output(), "%s\n", examples)
	}

	const (
		criticalUsage              = "set critical threshold"
		debugUsage                 = "set debug mode"
		debugDefaultVal            = false
		emptyOkUsage               = "empty data from graphite is ok"
		emptyOkDefaultVal          = false
		fromDefaultVal             = "-2min"
		fromUsage                  = "set from"
		graphiteURLDefaultVal      = ""
		graphiteURLUsage           = "graphite url"
		nameDefaultVal             = ""
		nameUsage                  = "set name to use in message and perfdata for target"
		maxUsage                   = "Maximum value for target"
		percentageDefaultVal       = false
		percentageUsage            = "calculate percentage and check against thresholds, uses -max or every second target as max for previous target"
		perfdataDefaultVal         = false
		perfdataUsage              = "show perfdata"
		sumDefaultVal              = false
		sumUsage                   = "summarize target datapoints and check against thresholds"
		targetDefaultVal           = ""
		targetUsage                = "comma-separated list of targets to check"
		targetNameRegexpDefaultVal = ""
		targetNameRegexpUsage      = "regex to show only the part of the target name, which matches"
		untilDefaultVal            = "now"
		untilUsage                 = "set until"
		warningUsage               = "set warning threshold"
	)

	flag.Float64Var(&checkGraphite.Critical, "critical", math.NaN(), criticalUsage)
	flag.Float64Var(&checkGraphite.Critical, "c", math.NaN(), criticalUsage)
	flag.BoolVar(&checkGraphite.Debug, "d", debugDefaultVal, debugUsage)
	flag.BoolVar(&checkGraphite.EmptyOk, "empty-ok", emptyOkDefaultVal, emptyOkUsage)
	flag.StringVar(&from, "from", fromDefaultVal, fromUsage)
	flag.StringVar(&from, "f", fromDefaultVal, fromUsage)
	flag.StringVar(&graphiteURL, "graphite-url", graphiteURLDefaultVal, graphiteURLUsage)
	flag.StringVar(&graphiteURL, "u", graphiteURLDefaultVal, graphiteURLUsage)
	flag.StringVar(&checkGraphite.Name, "name", nameDefaultVal, nameUsage)
	flag.Float64Var(&checkGraphite.Max, "max", math.NaN(), maxUsage)
	flag.Float64Var(&checkGraphite.Max, "m", math.NaN(), maxUsage)
	flag.BoolVar(&checkGraphite.Percentage, "percentage", percentageDefaultVal, percentageUsage)
	flag.BoolVar(&checkGraphite.Percentage, "p", percentageDefaultVal, percentageUsage)
	flag.BoolVar(&checkGraphite.Perfdata, "perfdata", perfdataDefaultVal, perfdataUsage)
	flag.BoolVar(&checkGraphite.SumFlag, "summarize", sumDefaultVal, sumUsage)
	flag.BoolVar(&checkGraphite.SumFlag, "s", sumDefaultVal, sumUsage)
	flag.Var(&targetFlag, "target", targetUsage)
	flag.Var(&targetFlag, "t", targetUsage)
	flag.StringVar(&checkGraphite.TargetNameRegexp, "target-name-regexp", targetNameRegexpDefaultVal, targetNameRegexpUsage)
	flag.StringVar(&checkGraphite.TargetNameRegexp, "r", targetNameRegexpDefaultVal, targetNameRegexpUsage)
	flag.DurationVar(&timeout, "timeout", 10, "timeout")
	flag.StringVar(&until, "until", untilDefaultVal, untilUsage)
	flag.Float64Var(&checkGraphite.Warning, "warning", math.NaN(), warningUsage)
	flag.Float64Var(&checkGraphite.Warning, "w", math.NaN(), warningUsage)

	return &checkGraphite
}

// Run parses command line arguments and run check
func (check *Check) Run() {
	flag.Parse()
	check.TargetFlag = &targetFlag

	defer check.Exit()

	data, err := graphite.GetTargets(graphiteURL, targetFlag, from, until, timeout)
	if err != nil {
		check.Error(err)
		return
	}

	check.runCheck(data)
}
